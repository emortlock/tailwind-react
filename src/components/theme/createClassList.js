import merge from 'lodash.merge'
import defaultTheme from './defaultTheme'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectTheme = {}) => {
  const theme = merge(defaultTheme, projectTheme)

  return [
    theme.radius,
    ...reduceMultiplePrefixes(theme.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(theme.container, 'max-w'),
    ...mapArray(theme.text.size.body, 'text'),
    ...mapArray(theme.text.size.title, 'text'),
    ...mapObject(theme.text.family, 'font'),
    ...reduceMultiplePrefixes(theme.brandColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(theme.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
