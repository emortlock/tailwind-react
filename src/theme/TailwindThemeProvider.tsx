import { Theme } from './TailwindTheme'

import React from 'react'
import merge from 'lodash.merge'

import { defaultTheme } from './defaultTheme'
import { TailwindTheme } from './TailwindTheme'

export const TailwindThemeProvider = ({
  theme = {},
  children,
}: {
  theme: Partial<Theme>
  children: React.ReactNode
}) => {
  const mergedTheme: Theme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}
