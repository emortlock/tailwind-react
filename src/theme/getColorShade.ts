const shades = ['100', '200', '300', '400', '500', '600', '700', '800', '900']

export const getColorShade = (
  color?: string | string[],
  offset: string | number = 1,
): string => {
  if (!color) return ''

  let currentColor: string | string[] = Array.isArray(color) ? color[0] : color

  if (offset === 0) return currentColor

  currentColor =
    currentColor === 'white' ? ['gray', '100'] : currentColor.split('-')
  let shadeOffset = offset

  if (color === 'white') {
    if (shadeOffset < 1) return color
    if (shadeOffset === 1) return 'gray-100'

    currentColor = ['gray', '100']
    if (typeof offset === 'number') shadeOffset = offset - 1
  }

  if (currentColor.length === 1) {
    currentColor.push('400')
  }

  if (typeof shadeOffset === 'string') {
    return `${currentColor[0]}-${shadeOffset}`
  }

  const shadeIndex = Math.min(
    Math.max(shades.indexOf(currentColor[1]) + shadeOffset, 0),
    shades.length - 1,
  )
  const newShade = shades[shadeIndex]

  return `${currentColor[0]}-${newShade}`
}
