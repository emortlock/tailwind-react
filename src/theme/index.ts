export * from './defaultTheme'
export * from './TailwindTheme'
export * from './TailwindThemeProvider'
export * from './withTheme'
export * from './useThemeValue'
export * from './getColorShade'
