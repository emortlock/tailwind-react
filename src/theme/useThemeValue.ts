import { tailwindPropToClassName } from 'tailwind-react-primitives'

export const useThemeValue = (
  prefix: string,
  value: unknown,
  userClassNames = '',
) => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
