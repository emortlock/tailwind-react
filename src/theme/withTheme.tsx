import type { Theme } from './TailwindTheme'

import React from 'react'

import { TailwindTheme } from './TailwindTheme'

export interface WithTheme {
  theme: Theme
}

export const withTheme = <P extends React.PropsWithChildren<{ theme: Theme }>>(
  Component: React.FC<P>,
) => {
  const WithTheme: React.FC<Omit<P, 'theme'>> = (props) => (
    <TailwindTheme.Consumer>
      {(theme) => {
        // @ts-ignore FIXME: Constraint error
        return <Component {...props} theme={theme} />
      }}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}
