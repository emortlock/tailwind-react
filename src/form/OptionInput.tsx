import type { TouchableProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'

import React from 'react'

import {
  RawTouchable as Touchable,
  RawFlex as Flex,
  RawBox as Box,
  Svg,
} from 'tailwind-react-primitives'
import { withTheme } from '../theme'

import { Label } from './Label'

export interface OptionInputProps<E extends HTMLElement = HTMLLabelElement>
  extends Omit<TouchableProps<E>, 'onChange'>,
    WithTheme {
  name: string
  value: string
  label: string
  hideLabel?: boolean
  defaultChecked?: boolean
  checked?: boolean
  onChange?: (e: React.ChangeEvent<HTMLElement>) => void
  checkbox?: boolean
  id?: string
}

export const RawOptionInput = <E extends HTMLElement = HTMLLabelElement>({
  defaultChecked,
  checked: externalChecked,
  theme,
  name,
  value,
  label,
  hideLabel,
  checkbox,
  id,
  onChange,
  ...rest
}: OptionInputProps<E>) => {
  const [checked, setChecked] = React.useState(
    defaultChecked || externalChecked,
  )
  const inputRef = React.useRef<HTMLInputElement | null>(null)

  React.useEffect(() => {
    if (externalChecked === undefined) return

    setChecked(externalChecked)
  }, [externalChecked])

  const handleChange = React.useCallback(
    (e: React.MouseEvent<E> | React.KeyboardEvent<E>) => {
      if (externalChecked === undefined) {
        setChecked((currentState) => (checkbox ? !currentState : true))
      }

      if (onChange && inputRef.current)
        onChange({
          ...e,
          target: inputRef.current,
        })

      e.preventDefault()
    },
    [],
  )

  return (
    <Touchable<E>
      as={Label}
      flex
      items="center"
      onTouch={handleChange}
      {...rest}
    >
      <Box<HTMLInputElement>
        as="input"
        id={id}
        visuallyHidden
        name={name}
        type={checkbox ? 'checkbox' : 'radio'}
        value={value}
        checked={checked}
        tabIndex={-1}
        innerRef={inputRef}
        onChange={() => {}}
      />
      <Flex
        items="center"
        justify="center"
        inlineBlock
        rounded={checkbox ? theme.radius : 'full'}
        h={4}
        w={4}
        border={[true, checked ? theme.brandColors.primary : false].filter(
          (prop) => !!prop,
        )}
        bg={checkbox && checked ? theme.brandColors.primary : undefined}
        m={{ r: theme.spacing.sm }}
      >
        {checked &&
          (checkbox ? (
            <Svg
              as="svg"
              xmlns="http://www.w3.org/2000/svg"
              viewBox="0 0 24 24"
              h={3}
              w={3}
              text="white"
              fill="current"
            >
              <path d="M0 0h24v24H0z" fill="none" />
              <path d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z" />
            </Svg>
          ) : (
            <Box
              inlineBlock
              rounded="full"
              h={2}
              w={2}
              bg={theme.brandColors.primary}
            />
          ))}
      </Flex>
      <Box inlineBlock visuallyHidden={hideLabel} leading="tight" font="normal">
        {label}
      </Box>
    </Touchable>
  )
}

export const OptionInput = withTheme<OptionInputProps>(RawOptionInput)
