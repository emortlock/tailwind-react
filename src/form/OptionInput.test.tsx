import React from 'react'
import { render, RenderResult } from '@testing-library/react'

import { RawOptionInput as OptionInput } from './OptionInput'
import { defaultTheme } from '../theme'

const setup = (
  testProps = {},
): [RenderResult, { props: Record<string, any> }] => {
  const props = Object.assign(
    { theme: defaultTheme, name: 'test', value: 'test-1', label: 'Test?' },
    testProps,
  )

  const wrapper = render(<OptionInput {...props} />)

  return [
    wrapper,
    {
      props,
    },
  ]
}

describe('OptionInput', () => {
  it('renders matching snapshot', () => {
    const [{ asFragment }] = setup({ defaultChecked: true })

    expect(asFragment()).toMatchSnapshot()
  })
})
