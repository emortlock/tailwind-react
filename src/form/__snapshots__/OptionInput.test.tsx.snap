// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionInput renders matching snapshot 1`] = `
<DocumentFragment>
  <label
    class="font-bold leading-normal inline-block mb-2 select-none cursor-pointer flex items-center focus:outline-none focus:shadow-outline"
    role="button"
    tabindex="0"
  >
    <input
      checked=""
      class="visually-hidden"
      name="test"
      tabindex="-1"
      type="radio"
      value="test-1"
    />
    <span
      class="inline-block flex items-center justify-center rounded-full h-4 w-4 border border-blue-400 mr-2"
    >
      <span
        class="inline-block rounded-full h-2 w-2 bg-blue-400"
      />
    </span>
    <span
      class="inline-block leading-tight font-normal"
    >
      Test?
    </span>
  </label>
</DocumentFragment>
`;
